internal/controllers/machineset/machineset_controller_status_test.go

```diff
@@ -0,0 +1,625 @@
+					Type:   clusterv1.MachineAvailableV1Beta2Condition,
+					Status: metav1.ConditionTrue,
+				}}}}},
+				{Status: clusterv1.MachineStatus{V1Beta2: &clusterv1.MachineV1Beta2Status{Conditions: []metav1.Condition{{
+					Type:   clusterv1.MachineAvailableV1Beta2Condition,
+					Status: metav1.ConditionFalse,
+				}}}}},
+				{Status: clusterv1.MachineStatus{V1Beta2: &clusterv1.MachineV1Beta2Status{Conditions: []metav1.Condition{{
+					Type:   clusterv1.MachineAvailableV1Beta2Condition,
+					Status: metav1.ConditionUnknown,
+				}}}}},
+			},
+			expectedStatus: &clusterv1.MachineSetV1Beta2Status{
+				ReadyReplicas:     ptr.To[int32](0),
+				AvailableReplicas: ptr.To[int32](1),
+				UpToDateReplicas:  ptr.To[int32](0),
+			},
+		},
+		{
+			name: "should count only up-to-date machines",
+			machines: []*clusterv1.Machine{
+				{Status: clusterv1.MachineStatus{V1Beta2: &clusterv1.MachineV1Beta2Status{Conditions: []metav1.Condition{{
+					Type:   clusterv1.MachineUpToDateV1Beta2Condition,
+					Status: metav1.ConditionTrue,
+				}}}}},
+				{Status: clusterv1.MachineStatus{V1Beta2: &clusterv1.MachineV1Beta2Status{Conditions: []metav1.Condition{{
+					Type:   clusterv1.MachineUpToDateV1Beta2Condition,
+					Status: metav1.ConditionFalse,
+				}}}}},
+				{Status: clusterv1.MachineStatus{V1Beta2: &clusterv1.MachineV1Beta2Status{Conditions: []metav1.Condition{{
+					Type:   clusterv1.MachineUpToDateV1Beta2Condition,
+					Status: metav1.ConditionUnknown,
+				}}}}},
+			},
+			expectedStatus: &clusterv1.MachineSetV1Beta2Status{
+				ReadyReplicas:     ptr.To[int32](0),
+				AvailableReplicas: ptr.To[int32](0),
+				UpToDateReplicas:  ptr.To[int32](1),
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+			ms := &clusterv1.MachineSet{}
+			setReplicas(ctx, ms, tt.machines, tt.machines != nil)
+			g.Expect(ms.Status.V1Beta2).To(BeEquivalentTo(tt.expectedStatus))
+		})
+	}
+}
+
+func Test_setScalingUpCondition(t *testing.T) {
+	defaultMachineSet := &clusterv1.MachineSet{
+		Spec: clusterv1.MachineSetSpec{
+			Replicas: ptr.To[int32](0),
+			Template: clusterv1.MachineTemplateSpec{
+				Spec: clusterv1.MachineSpec{
+					Bootstrap: clusterv1.Bootstrap{
+						ConfigRef: &corev1.ObjectReference{
+							Kind:      "KubeadmBootstrapTemplate",
+							Namespace: "some-namespace",
+							Name:      "some-name",
+						},
+					},
+					InfrastructureRef: corev1.ObjectReference{
+						Kind:      "DockerMachineTemplate",
+						Namespace: "some-namespace",
+						Name:      "some-name",
+					},
+				},
+			},
+		},
+	}
+
+	scalingUpMachineSetWith3Replicas := defaultMachineSet.DeepCopy()
+	scalingUpMachineSetWith3Replicas.Spec.Replicas = ptr.To[int32](3)
+
+	deletingMachineSetWith3Replicas := scalingUpMachineSetWith3Replicas.DeepCopy()
+	deletingMachineSetWith3Replicas.DeletionTimestamp = ptr.To(metav1.Now())
+	deletingMachineSetWith3Replicas.Status.Replicas = 2
```

Comment:
```suggestion
	scalingUpMachineSetWith3Replicas := defaultMachineSet.DeepCopy()
	scalingUpMachineSetWith3Replicas.Spec.Replicas = ptr.To[int32](3)
	scalingUpMachineSetWith3Replicas.Status.Replicas = 0

	deletingMachineSetWith3Replicas := defaultMachineSet.DeepCopy()
	deletingMachineSetWith3Replicas.DeletionTimestamp = ptr.To(metav1.Now())
	deletingMachineSetWith3Replicas.Spec.Replicas = ptr.To[int32](3)
	deletingMachineSetWith3Replicas.Status.Replicas = 2
```
nit: Maybe let's make this slightly more explicit